File: samples/Rust/binary-search.rs
fn binary_search(haystack: ~[int], needle: int) -> int {
    let mut low  = 0;
    let mut high = haystack.len() as int - 1;

    if high == 0 { return -1 }

    while low <= high {
        // avoid overflow
        let mid = low + (high - low) / 2;

        if      haystack[mid] > needle { high = mid - 1 }
        else if haystack[mid] < needle { low  = mid + 1 }
        else                           { return mid     }
    }

    return -1;
}

File: samples/Scala/binary-search.scala
def binarySearch[A <% Ordered[A]](a: IndexedSeq[A], v: A) = {
  def recurse(low: Int, high: Int): Option[Int] = (low + high) / 2 match {
    case _ if high < low => None
    case mid if a(mid) > v => recurse(low, mid - 1)
    case mid if a(mid) < v => recurse(mid + 1, high)
    case mid => Some(mid)
  }
  recurse(0, a.size - 1)
}
